test/suites/kubernetesupgrade/suite_test.go

```diff
@@ -0,0 +1,118 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetesupgrade_test
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/Azure/karpenter-provider-azure/pkg/apis/v1alpha2"
+	"github.com/Azure/karpenter-provider-azure/test/pkg/environment/azure"
+
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+
+	coretest "sigs.k8s.io/karpenter/pkg/test"
+)
+
+var env *azure.Environment
+var nodeClass *v1alpha2.AKSNodeClass
+var nodePool *karpv1.NodePool
+
+func TestDrift(t *testing.T) {
+	RegisterFailHandler(Fail)
+	BeforeSuite(func() {
+		env = azure.NewEnvironment(t)
+	})
+	RunSpecs(t, "KubernetesUpgrade")
+}
+
+var _ = BeforeEach(func() {
+	env.BeforeEach()
+	nodeClass = env.DefaultAKSNodeClass()
+	nodePool = env.DefaultNodePool(nodeClass)
+})
+var _ = AfterEach(func() { env.Cleanup() })
+var _ = AfterEach(func() { env.AfterEach() })
+
+var _ = Describe("KubernetesUpgrade", func() {
+	// NOTE: This Describe has only one test, and no BeforeEach, as this test preforms modifications to the managed cluster in a way that can not be undone,
+	// meaning the environment can not be reset for any additional tests.
+	//
+	// This is actually a test of Drift. However, testing kubernetes upgrade has to be isolated due to its modification of the managed cluster environment.
+	It("should drift and replace all nodes when the cluster's kubernetes version upgrades", func() {
+		numPods := 1
+		// Add pods with a do-not-disrupt annotation so that we can check node metadata before we disrupt
+		dep := coretest.Deployment(coretest.DeploymentOptions{
+			Replicas: int32(numPods),
+			PodOptions: coretest.PodOptions{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": "my-app",
+					},
+					Annotations: map[string]string{
+						karpv1.DoNotDisruptAnnotationKey: "true",
+					},
+				},
+				TerminationGracePeriodSeconds: lo.ToPtr[int64](0),
+			},
+		})
+		selector := labels.SelectorFromSet(dep.Spec.Selector.MatchLabels)
+
+		env.ExpectCreated(dep, nodeClass, nodePool)
+		pod := env.EventuallyExpectHealthyPodCount(selector, numPods)[0]
+		nodeClaim := env.EventuallyExpectCreatedNodeClaimCount("==", 1)[0]
+		node := env.ExpectCreatedNodeCount("==", 1)[0]
+
+		By("getting latest available kubernetes version upgrade")
+		availableKubernetesUpgrades := env.ExpectSuccessfulGetOfAvailableKubernetesVersionUpgradesForManagedCluster()
+		sort.Slice(availableKubernetesUpgrades, func(i, j int) bool {
+			return *availableKubernetesUpgrades[i].KubernetesVersion < *availableKubernetesUpgrades[j].KubernetesVersion
+		})
+		// Get the latest available version
+		kubernetesUpgradeVersion := *availableKubernetesUpgrades[len(availableKubernetesUpgrades)-1].KubernetesVersion
```

Comment:
Maybe `lo.MaxBy`?

```suggestion
		kubernetesUpgradeVersion := lo.MaxBy(availableKubernetesUpgrades, func(a, b *containerservice.ManagedClusterPoolUpgradeProfileUpgradesItem) bool {
			return *a.KubernetesVersion < *b.KubernetesVersion
		}).KubernetesVersion
```

Reply:
Added, but actually looks like MaxBy comparision func is reverse, so switched to `>`

Reply:
Also, shouldn't this technically be semver comparison?

Reply:
So, I was considering if it should be. I mean in theory if we start using prebuilds of k8s, etc doing a string comparison might not hold up (I'm not sure on those comparisons). Otherwise, I believe the mapping of string comparison actually lines up with the versioning comparison, so I decided to go with this shortcut in the tests. 

I can just make the change though. Wouldn't make a shortcut like this in actual code, so probably shouldn't in testing either.